frontend/views/containers/user-settings/AppLogs.vue

```diff
@@ -0,0 +1,177 @@
+<template lang='pug'>
+  .settings-container
+    section.card
+      .c-header
+        fieldset.c-filters
+          legend.sr-only Visible logs
+          label.checkbox
+            input.input(type='checkbox' name='filter' v-model='form.filter' value='error')
+            i18n Errors
+          label.checkbox
+            input.input(type='checkbox' name='filter' v-model='form.filter' value='warn')
+            i18n Warnings
+          label.checkbox
+            input.input(type='checkbox' name='filter' v-model='form.filter' value='debug')
+            i18n Debug
+          label.checkbox
+            input.input(type='checkbox' name='filter' v-model='form.filter' value='info')
+            i18n Info
+          label.checkbox
+            input.input(type='checkbox' name='filter' v-model='form.filter' value='log')
+            i18n Log
+
+        button.is-small.c-download(@click='downloadLogs')
+          i.icon-download.is-prefix.c-icon
+          i18n.hide-touch Download
+          i18n.hide-desktop Share
+        a(ref='linkDownload' hidden)
+
+      textarea.textarea.c-logs(ref='textarea' rows='12' readonly)
+        | {{ prettyLogs }}
+
+      i18n.link(tag='a' @click='openTroubleshooting') Troubleshooting
+</template>
+
+<script>
+import { mapState, mapMutations, mapActions } from 'vuex'
+import sbp from '~/shared/sbp.js'
+import { CAPTURED_LOGS, SET_APP_LOGS_FILTER } from '@utils/events.js'
+import { downloadLogs } from '@model/captureLogs.js'
+export default {
+  name: 'AppLogs',
+  components: {},
+  data () {
+    return {
+      form: {
+        filter: []
+      },
+      ephemeral: {
+        logs: []
+      }
+    }
+  },
+  created () {
+    this.form.filter = this.appLogsFilter
+    sbp('okTurtles.events/on', CAPTURED_LOGS, this.addLog)
+
+    const logs = []
+    let lastEntry = localStorage.getItem('giConsole/lastEntry')
+    do {
+      const entry = JSON.parse(localStorage.getItem(`giConsole/${lastEntry}`))
+      if (!entry) break
+      logs.push(entry)
+      lastEntry = entry.prev
+    } while (lastEntry)
+    this.ephemeral.logs = logs.reverse()
```

Comment:
Instead of `logs.push(entry)`, do:

```js
this.ephemeral.logs.unshift(entry)
```

This way no `.reverse()` is needed! :D

Reply:
Initially, I did that, but that's not a fast approach. Why? Because `.push` is o(1) and `.unshift` is o(n).

So this current approach, each iteration of `do... while` is O(1) and the last step is o(n) - `.reverse` is o(n) too.

Reply:
I didn't know `.unshift` was `O(n)`... OK, that makes sense 👍 

(There are some optimizations JS engines could do to make it `O(1)`, but you're right it's not guaranteed that they do them...)